pkg/ccl/jwtauthccl/authentication_jwt.go

```diff
@@ -132,22 +142,44 @@ func (authenticator *jwtAuthenticator) ValidateJWTLogin(
 
 	telemetry.Inc(beginAuthUseCounter)
 
-	parsedToken, err := jwt.Parse(tokenBytes, jwt.WithKeySet(authenticator.mu.conf.jwks), jwt.WithValidate(true), jwt.InferAlgorithmFromKey(true))
+	// Just parse the token to check the format is valid and issuer is present.
+	// The token will be parsed again later to actually verify the signature.
+	unverifiedToken, err := jwt.Parse(tokenBytes)
 	if err != nil {
 		return errors.Newf("JWT authentication: invalid token")
 	}
 
+	// Check for issuer match against configured issuers.
+	issuerUrl := ""
 	issuerMatch := false
 	for _, issuer := range authenticator.mu.conf.issuers {
-		if issuer == parsedToken.Issuer() {
+		if issuer == unverifiedToken.Issuer() {
 			issuerMatch = true
+			issuerUrl = issuer
 			break
 		}
 	}
 	if !issuerMatch {
 		return errors.WithDetailf(
 			errors.Newf("JWT authentication: invalid issuer"),
-			"token issued by %s", parsedToken.Issuer())
+			"token issued by %s", unverifiedToken.Issuer())
+	}
+
+	var jwkSet jwk.Set
+	// If auto-fetch is enabled, fetch the JWKS remotely from the issuer's well known jwks url.
+	if authenticator.mu.conf.jwksAutoFetchEnabled {
+		jwkSet, err = remoteFetchJWKS(ctx, issuerUrl)
+		if err != nil {
+			return errors.Newf("JWT authentication: unable to validate token")
```

Comment:
The original error is lost in these handlers. I know we want to avoid leaking sensitive information, but this will make debugging this feature more challenging. Is there a reason we didn't wrap the `err` here and elsewhere?

Reply:
This is the discussion that lead to removal of logging lines. I agree that in some cases it might make it difficult to debug the exact error, but it is inline with other error handlings in this file. I am open to consider an alternate solution and include if available. https://reviewable.io/reviews/cockroachdb/cockroach/117054#-Nne7jnh5oeb9srWuZw2

Reply:
in some other places (e.g. password checking), we use the `AuthConn.LogAuthFailed()` helper so the detailed failure can be logged internally, and we show a more generic message to the user who failed auth.

https://github.com/cockroachdb/cockroach/blob/128b3967d19406eba6c706c1b0063375194be55e/pkg/sql/pgwire/auth.go#L141

i don't think we have access to that logger in this part of the code. refactoring the code is possible, but i'd say can be done separately from this backport